### ISSUE TITLE
Compiler bug: "too many arguments in call" on MacOS

### ISSUE THREAD

**Issue by srackham** [vlang/v#17914]
### Describe the bug

Errors compiling C code emitted by V: `error: too many arguments in call`. The error only manifests in MacOS (Linux and Windows pass).

### Expected Behavior

No C compile errors.

### Current Behavior

I don't have access to a macOS so I can only report the Github Actions error.
Both Linux and MacOS run the same Github actions. Here are the Github actions test run logs:

https://github.com/srackham/v-rimu/actions/runs/4603691234/jobs/8226161609

Here's an error from the above logs:

```
FAIL  [ 3/17]   731.845 ms /Users/runner/work/v-rimu/v-rimu/v-rimu/delimitedblocks/delimitedblocks_test.v
==================
/tmp/v_501/tsession_117b51600_168750688/delimitedblocks_test.12765202679951327310.tmp.c:3854:55: error: too many arguments in call to 'anon_fn_array_string__bool_str' [-Werror]
        string _t3 = anon_fn_array_string__bool_str(it.verify);
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~          ^
/tmp/v_501/tsession_117b51600_168750688/delimitedblocks_test.12765202679951327310.tmp.c:3855:95: error: too many arguments in call to 'anon_fn_array_string_delimitedblocks__definition__string_str' [-Werror]
        string _t4 = anon_fn_array_string_delimitedblocks__definition__string_str(it.delimiter_filter);
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                    ^
/tmp/v_501/tsession_117b51600_168750688/delimitedblocks_test.12765202679951327310.tmp.c:3856:91: error: too many arguments in call to 'anon_fn_string_array_string_expansion__options__string_str' [-Werror]
        string _t5 = anon_fn_string_array_string_expansion__options__string_str(it.content_filter);
                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                  ^
3 errors generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.
```

The errors center around compiled C functions such as `anon_fn_array_string__bool_str`, (compiled on Linux):

```
static string anon_fn_array_string__bool_str(); // auto
  :
static string anon_fn_array_string__bool_str() { return _SLIT("fn ([]string) bool");}
  :
	string _t3 = anon_fn_array_string__bool_str(it.verify);
```

### Reproduction Steps

I don't have access to macOS so haven't been able to debug this problem and narrow the issue down, so unfortunately there's not a lot to go on.

### V version

V 0.3.3 cd629cf

### Environment details (OS name and version, etc.)

macOS 12.6.4 21G526 running on Github

**Comment by JalonSolov** [comment]
So this is failing on your latest commit, #[dca8795](https://github.com/srackham/v-rimu/commit/dca8795596d6942e758510dd57470369bb9c52e7), right?

**Comment by srackham** [comment]
That's correct, so I guess it is related to the changes to the `parse()` function in the `blockattributes.v` file. Either that or a V regression in the last 2 months but it does not look like you can rerun old Github workflows to check.

**Comment by JalonSolov** [comment]
No, once they "age out", all you can do is push a change to get them to run again... but then you're not running against the same commit.

If the code was in a PR, you could close then reopen the PR to get them to run again on the same commit.

**Comment by spytheman** [comment]
That is not a problem in V, at least not directly.

In your Makefile, you have:
```makefile
test: fmt build-rimuv
	v -cstrict -enable-globals test .
```

On macos, clang is 14.0.0 or even 15.0, which is even stricter, so if you pass `-cstrict`, all C warnings that it detects will get converted to C errors instead.

The following works (it progresses further):
`brew install pcre2`
`v -cflags -L/opt/homebrew/lib  -enable-globals test .`

This also compiles, even with `-cstrict`:
`v -L/opt/homebrew/lib -cstrict -enable-globals -o bin/rimuv cmd/rimuv/rimuv.v`

**Comment by spytheman** [comment]
After the last command, that compiled bin/rimuv, even the test works now (previously it had some failing assertions, probably it required the binary to be present 🤔 ):
<img width="987" alt="image" src="https://user-images.githubusercontent.com/26967/230954344-2e2f7713-599c-458c-93dd-7d02df39a67e.png">

**Comment by spytheman** [comment]
@srackham , I suggest adding `#flag darwin -L /opt/homebrew/lib` after the already existing `~/.vmodules/srackham/pcre2/pcre2.c.v:10:#flag darwin -I /opt/homebrew/include/`

**Comment by spytheman** [comment]
Another thing to note is that `fn init(){` is called by V itself automatically, as part of module initialization, while in `delimitedblocks_test.v` you call it manually also several times, which may or may not be a problem.

**Comment by srackham** [comment]
@spytheman. You nailed it! CI tests pass now. Thank you very much.

- https://github.com/srackham/v-rimu/commit/62688ecb6b68d43c4ea1e68db43f3555d55173a6
- https://github.com/srackham/pcre2/commit/358185ba8aaa4a804279875dc4525b9a6e6ca71f

**Comment by srackham** [comment]
> Another thing to note is that `fn init(){` is called by V itself automatically (as part of _vinit, which is executed before all other tests or your main function for normal programs), as part of module initialization, while in `delimitedblocks_test.v` you call it manually also several times, which may or may not be a problem.

Thanks for picking that up. It doesn't cause problems but calling `init` directly is confusing, tests should call `initialize` instead so that the intention is clearer (https://github.com/srackham/v-rimu/commit/2ec12e060ce2ecc1d4244f546275adc9f1940f1c).

**Comment by srackham** [comment]
I guess we're still left with the compiler `too many arguments in call` error if `-cstrict` is used, which is caused by compiler generating C function calls with incorrect function signatures (diagnosed by @spytheman here https://github.com/vlang/v/issues/17914#issuecomment-1502091710)

**Comment by spytheman** [comment]
> I guess we're still left with the compiler too many arguments in call error if -cstrict is used

Yes, it is a cgen issue, but I have not made a smaller reproduction yet (I got distracted yesterday). I will do so asap.

**Comment by felipensp** [comment]
```V
fn main() {
	a := fn (a string) int {
		return 1
	}
	println(a)
}
```

```
v -no-retry-compilation  -g -cstrict -cc clang run bug.v
/home/felipe/github/v/bug.v:6:64: error: too many arguments in call to 'anon_fn_70a6b7e29084d475_string__int_37_str' [-Werror]
        println(anon_fn_70a6b7e29084d475_string__int_37_str((voidptr)a));
                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~           ^
/home/felipe/github/v/bug.v:6:53: error: passing arguments to 'anon_fn_70a6b7e29084d475_string__int_37_str' without a prototype is deprecated in all versions of C and is not supported in C2x [-Werror,-Wdeprecated-non-prototype]
        println(anon_fn_70a6b7e29084d475_string__int_37_str((voidptr)a));
                                                           ^
2 errors generated.

```
